src/langchain_google_memorystore_redis/chat_message_history.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import redis
+from typing import List, Optional
+
+from langchain_core.chat_history import BaseChatMessageHistory
+from langchain_core.messages import (
+    BaseMessage,
+    message_to_dict,
+    messages_from_dict,
+)
+
+
+class MemorystoreChatMessageHistory(BaseChatMessageHistory):
+    """Chat message history stored in a Cloud Memorystore for Redis database."""
+
+    def __init__(
+        self,
+        client: redis.Redis,
+        session_id: str,
+        ttl: Optional[int] = None,
+    ):
+        """Initializes the chat message history for Memorystore for Redis.
+
+        Args:
+            client: A redis.Redis client object.
+            session_id: A string that uniquely identifies the chat history.
+            ttl: Specifies the time in seconds after which the session will
+                expire and be eliminated from the Redis instance since the most
+                recent message is added.
+        """
+
+        self.redis = client
+        self.key = session_id
+        self.ttl = ttl
+
+        all_elements = self.redis.lrange(self.key, 0, -1)
```

Comment:
should we try/except Redis errors?

Reply:
Redis exception is directly thrown out to the client application if it encounters any issue. This aligns with the community practice like: https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/chat_message_histories/elasticsearch.py#L106-L116. They directly throw out exceptions if any setup is invalid.

In our example, if the user provides an invalid address, they will see direct exception from redis-py:

```
>>> c2 = MemorystoreChatMessageHistory(redis.from_url('redis://localhost:6378'), 'chat1')
Traceback (most recent call last):
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/connection.py", line 264, in connect
    sock = self.retry.call_with_retry(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/retry.py", line 46, in call_with_retry
    return do()
           ^^^^
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/connection.py", line 265, in <lambda>
    lambda: self._connect(), lambda error: self.disconnect(error)
            ^^^^^^^^^^^^^^^
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/connection.py", line 627, in _connect
    raise err
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/connection.py", line 615, in _connect
    sock.connect(socket_address)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/google/home/craigchi/venv/langchain/langchain-google-memorystore-redis-python/src/langchain_google_memorystore_redis/chat_message_history.py", line 50, in __init__
    all_elements = self.redis.lrange(self.key, 0, -1)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/commands/core.py", line 2754, in lrange
    return self.execute_command("LRANGE", name, start, end)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/client.py", line 533, in execute_command
    conn = self.connection or pool.get_connection(command_name, **options)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/connection.py", line 1086, in get_connection
    connection.connect()
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/connection.py", line 270, in connect
    raise ConnectionError(self._error_message(e))
redis.exceptions.ConnectionError: Error 111 connecting to localhost:6378. Connection refused.
```